File: experimental/ImPlay/Implay.Core/Services/StartupLogger.cs
using System.Text;

namespace ImPlay.Core.Services;

public static class StartupLogger
{
    private static readonly object LockObj = new();
    private static string? _logPath;

    public static string LogPath => _logPath ??= Initialize();

    public static string Initialize()
    {
        lock (LockObj)
        {
            var configDir = PathHelper.GetConfigDir();
            var path = Path.Combine(configDir, "startup.log");
            _logPath = path;

            var header = new StringBuilder();
            header.AppendLine(new string('-', 72));
            header.AppendLine($"=== ImPlay startup {DateTimeOffset.Now:O} ===");
            header.AppendLine($"BaseDir={AppDomain.CurrentDomain.BaseDirectory}");
            header.AppendLine();

            File.AppendAllText(path, header.ToString(), Encoding.UTF8);
            return path;
        }
    }

    public static void Log(string message)
    {
        lock (LockObj)
        {
            var line = $"[{DateTimeOffset.Now:O}] {message}\n";
            try
            {
                File.AppendAllText(LogPath, line, Encoding.UTF8);
            }
            catch
            {
                // Ignore logging failures
            }
        }
    }

    public static void LogException(string context, Exception ex)
    {
        Log($"{context} failed: {ex.Message}\n{ex.StackTrace}");
    }
}

File: experimental/implayfsharpavalonia/Core/StartupLogger.fs
namespace ImPlay.Core.Services

open System
open System.IO
open System.Text
open ImPlay.Core.Helpers

module StartupLogger =
    let private lockObj = obj()
    let mutable private _logPath : string option = None

    let initialize() =
        lock lockObj (fun () ->
            let configDir = PathHelper.GetConfigDir()
            let path = Path.Combine(configDir, "startup.log")
            _logPath <- Some path
            
            let sb = StringBuilder()
            sb.AppendLine(String('-', 72)) |> ignore
            sb.AppendLine($"=== ImPlay startup {DateTimeOffset.Now:O} ===") |> ignore
            sb.AppendLine($"BaseDir={AppDomain.CurrentDomain.BaseDirectory}") |> ignore
            sb.AppendLine() |> ignore
            
            File.AppendAllText(path, sb.ToString(), Encoding.UTF8)
            path
        )

    let logPath() =
        match _logPath with
        | Some p -> p
        | None -> initialize()

    let log (message: string) =
        lock lockObj (fun () ->
            let line = $"[{DateTimeOffset.Now:O}] {message}\n"
            try
                File.AppendAllText(logPath(), line, Encoding.UTF8)
            with _ -> ()
        )

    let logException (context: string) (ex: Exception) =
        log $"{context} failed: {ex.Message}\n{ex.StackTrace}"
